File: python/find_all_anagrams_in_a_string.py
class Solution:
    def findAnagrams(self, s: str, p: str) -> list[int]:
        if len(p) > len(s):
            return []

        matches: int = 0
        result: list[int] = []
        s_count: list[int] = [0] * 26
        p_count: list[int] = [0] * 26

        for i in range(len(p)):
            s_count[ord(s[i]) - ord("a")] += 1
            p_count[ord(p[i]) - ord("a")] += 1

        for i in range(26):
            matches += 1 if s_count[i] == p_count[i] else 0

        l: int = 0
        for r in range(len(p), len(s)):
            if matches == 26:
                result.append(l)

            rightIndex: int = ord(s[r]) - ord("a")
            s_count[rightIndex] += 1
            if s_count[rightIndex] == p_count[rightIndex]:
                matches += 1
            elif s_count[rightIndex] - 1 == p_count[rightIndex]:
                matches -= 1

            leftIndex: int = ord(s[l]) - ord("a")
            s_count[leftIndex] -= 1
            if s_count[leftIndex] == p_count[leftIndex]:
                matches += 1
            elif s_count[leftIndex] + 1 == p_count[leftIndex]:
                matches -= 1

            l += 1

        result.append(l) if matches == 26 else None

        return result


def test_find_all_anagrams_in_a_string_case_1():
    # arrange
    s = "cbaebabacd"
    p = "abc"
    expected = [0, 6]

    # act
    solution = Solution()
    actual = solution.findAnagrams(s, p)

    # assert
    assert actual == expected


def test_find_all_anagrams_in_a_string_case_2():
    # arrange
    s = "abab"
    p = "ab"
    expected = [0, 1, 2]

    # act
    solution = Solution()
    actual = solution.findAnagrams(s, p)

    # assert
    assert actual == expected

File: rust/src/code/find_all_anagrams_in_a_string.rs
struct Solution {}

impl Solution {
    pub fn find_anagrams(s: String, p: String) -> Vec<i32> {
        if s.len() < p.len() {
            return Vec::<i32>::new();
        }

        let mut matches = 0;
        let mut result: Vec<i32> = Vec::new();
        let mut s_count = [0; 26];
        let mut p_count = [0; 26];
        let s_chars: Vec<char> = s.chars().collect();
        let p_chars: Vec<char> = p.chars().collect();

        for i in 0..p_chars.len() {
            p_count[(p_chars[i] as u8 - 'a' as u8) as usize] += 1;
            s_count[(s_chars[i] as u8 - 'a' as u8) as usize] += 1;
        }

        for i in 0..26 {
            if p_count[i] == s_count[i] {
                matches += 1;
            }
        }

        let mut s: usize = 0;
        for e in p_chars.len()..s_chars.len() {
            if matches == 26 {
                result.push(s as i32);
            }

            let index = (s_chars[e] as u8 - 'a' as u8) as usize;
            s_count[index] += 1;
            if p_count[index] == s_count[index] {
                matches += 1;
            } else if p_count[index] == s_count[index] - 1 {
                matches -= 1;
            }

            let index = (s_chars[s] as u8 - 'a' as u8) as usize;
            s_count[index] -= 1;
            if p_count[index] == s_count[index] {
                matches += 1;
            } else if p_count[index] == s_count[index] + 1 {
                matches -= 1;
            }

            s += 1;
        }

        if matches == 26 {
            result.push(s as i32);
        }

        result
    }
}

#[cfg(test)]
mod test {
    use super::Solution;

    #[test]
    fn find_anagrams_case_1() {
        // arrange
        let s = "cbaebabacd".to_string();
        let p = "abc".to_string();
        let expected = vec![0, 6];

        // act
        let actual = Solution::find_anagrams(s, p);

        // assert
        assert_eq!(expected, actual);
    }

    #[test]
    fn find_anagrams_case_2() {
        // arrange
        let s = "abab".to_string();
        let p = "ab".to_string();
        let expected = vec![0, 1, 2];

        // act
        let actual = Solution::find_anagrams(s, p);

        // assert
        assert_eq!(expected, actual);
    }
}
